Fix type layout when Cross OS compiling

sdmaclea opened the pull request:
This is a result of an audit of mismatching types in the Linux DAC vs the Windows-Linux X DAC. It fixes the _real_ differences identified by the tool.

sdmaclea committed:
Fix cache line size to be a TARGET property

sdmaclea committed:
Fix CCompRC cross OS DAC layout

sdmaclea committed:
Fix cross OS InlineSArray packing

sdmaclea committed:
Add DAC_ALIGNAS()

Align first member of various classes which have shown
inconsistent layout during cross compilation.

sdmaclea committed:
Fix RegMeta cross os layout

jkotas commented on src/coreclr/src/inc/daccess.h:
```diff
@@ -2458,4 +2458,16 @@ typedef DPTR(PTR_PCODE) PTR_PTR_PCODE;
 
+// For cross compilation, controlling type layout is important
+// We add a simple macro here which defines DAC_ALIGNAS to the C++11 alignas operator
+// This helps force the algnment of the next member
```
```suggestion
// This helps force the alignment of the next member
```

jkotas commented:
How are we going to make sure that there macros are in all the right places? Their placement seems completely ad-hoc to me.

jkotas commented on src/coreclr/src/vm/loaderallocator.hpp:
```diff
@@ -151,6 +151,8 @@ class LoaderAllocator
     // Objects that are allocated frequently should be allocated into the HighFreq heap for
     // better page management
+
+    DAC_ALIGNAS(8) // Align the first member to an 8 byte boundary.  Windows does this by default, force Linux to match.
```
Is this correct even on 32-bit platforms?

sdmaclea replied:
This was added because with the x86 Windows compiler this member was aligned to a 8 byte boundary.  Based on all the diff I looked at, I am inferring that the first member is aligned to the `alignof` the class.  `DAC_ALIGNAS(LoaderAllocator)` would be semantically correct, but I am not sure if it is supported (because the type is incomplete when I need to write the line).

sdmaclea replied:
On Windows, the class and first member is aligned to 8-bytes even on x86/arm because
- The `uint64_t m_nLoaderAllocator` member
- Windows packing is required to be set to 8. Linux must be configured to match.

jkotas replied:
Can this be `DAC_ALIGNAS(m_nLoaderAllocator)`? Or at lest refer to `m_nLoaderAllocator` in the comment?

sdmaclea commented:
>Their placement seems completely ad-hoc to me.

I saw two issues 
+ Base class required the compiler to pad the `class` to make `sizeof(BaseClass)%alignof(BaseClass)`.  This was fairly common.  In this case, the gcc/clang compiler would sneak the first member on top of the base class padding.
+ Virtual class with `alignof(VirtualClass) > alignof(void*)`.  This occurred in exactly one place. The Windows compiler would force the alignment of the first member to the alignment of the class.

Didn't want to litter lots of code with these layout corrections. So they are currently only in the places I saw an issue when looking at the arm, arm64, and x64 cross OS layout.

I tried to make the macros as robust as possible. They are placed immediately before the declaration of the  first member, and the alignment requirement are in terms of the base class.

> How are we going to make sure that there macros are in all the right places? 

I am thinking it will be maintained through SOS testing.

I wrote a PDB Dwarf layout compare tool for this exercise. I could eventually integrate into the `dotnet/diagnostic` test infra. I was thinking it was the most important piece of X DAC test infra. So far I have been alone...

sdmaclea committed:
Fix typo

jkotas commented on src/coreclr/src/inc/utilcode.h:
```diff
@@ -775,11 +775,11 @@ class CCompRC
 
     LPCWSTR m_pResourceFile;
-#ifdef HOST_UNIX
+#ifdef TARGET_UNIX
```
This is not DACized, so it should not be needed. 

Also, this type is designed to run on HOST not TARGET.

jkotas replied:
Actually, it looks like all this is dead code, so it would be best to delete it.

jkotas commented on src/coreclr/src/vm/stubgen.h:
```diff
@@ -261,6 +261,7 @@ class FunctionSigBuilder : protected StubSigBuilder
     }
 
 protected:
+    DAC_ALIGNAS(StubSigBuilder) // Align the first member to the alignment of the base class
```
This type is not DACized. This should not be needed. Would it be better to ifdef it out for DAC?

sdmaclea committed:
Remove dead code

sdmaclea committed:
Remove FunctionSigBuilder et. al from DAC

Remove FunctionSigBuilder and classes which depend on its
definition from the DAC.

Remove the FunctionSigBuilder layout correction to make it
cross platform stable.

jkotas commented on src/coreclr/src/vm/stubgen.h:
```diff
@@ -205,6 +205,8 @@ class LocalSigBuilder : protected StubSigBuilder
 
 };  // class LocalSigBuilder
 
+#ifndef DACCESS_COMPILE
```
Can `LocalSigBuilder` and all of the above it be inside this ifndef as well?

sdmaclea committed:
Remove more stubgen classes from DAC

noahfalk commented:
> >How are we going to make sure that there macros are in all the right places?
> 
> I am thinking it will be maintained through SOS testing.
> I wrote a PDB Dwarf layout compare tool for this exercise. I could eventually integrate into the dotnet/diagnostic test infra. I was thinking it was the most important piece of X DAC test infra. So far I have been alone...

I had to step out for another meeting during part of the meeting where testing was discussed so perhaps this came up then? What were the objections? (I am infering the 'alone' comment meant there were some?)

noahfalk reviewed (approved):
Code changes seem fine to me. Similar to Jan I think the main question it raises is ensuring that if another case pops up in the future we have a way to find it.

sdmaclea commented:
> I had to step out for another meeting during part of the meeting where testing was discussed so perhaps this came up then? What were the objections? (I am infering the 'alone' comment meant there were some?)

This was referring to an old conversation @Noahfalk & I had in Jan/Dec.  At the time you had suggested this could be a one off exercise. In my mind, the type layout was the ugliest part of this plan. When a new compiler just break everything and we don't know how to fix it.

I'll plan on eventually adding type layout of testing somewhere somehow...

sdmaclea committed:
Improve comment

noahfalk commented:
> This was referring to an old conversation @noahfalk & I had in Jan/Dec. At the time you had suggested this could be a one off exercise

I don't remember exactly what I said or the context, but my apologies if it created confusion. Let me attempt to restate my position:
1. We should have some form of ongoing regression testing to catch newly introduced issues. 
2. The form of that ongoing testing is flexible. It isn't required to be an explicit comparison of type layouts but it should at least detect the issues that a type layout mismatch would cause. For example running a bunch of SOS tests and confirming all the commands give expected results would probably be one way to do that.

sdmaclea commented:
@noahfalk I guess I understood that position at the time. I think you had a strong preference to full end to end testing over explicit layout of types.  I think I had the opposite bias...  It probably led me to misunderstand your opinion and distort it as time passed.

The explicit layout comparison has a nice advantage of getting to root cause.
End to end testing provides better overall functional coverage at the expense of longer debugging.  
Both would be best...

noahfalk commented:
> The explicit layout comparison has a nice advantage of getting to root cause.
End to end testing provides better overall functional coverage at the expense of longer debugging.
Both would be best...

That sounds accurate : ) 'Both' is of course more engineering work, but if you think the extra labor pays for itself with faster issue resolutions down the road I am happy to support it.
